### ISSUE TITLE
Add `request_multi`

### ISSUE THREAD

**Issue by autodidaddict** [nats-io/nats.ex#117]
The other NATS clients all support the ability to make a request and then gather a list of replies within a given timeout period. While this is possible to do manually using the existing functions, it's awkward and error prone and we should have an officially supported way of doing this. The scatter-gather pattern is pretty commonplace when using NATS.

Rust reference: [https://docs.rs/nats/0.16.0/nats/struct.Connection.html#method.request_multi](https://docs.rs/nats/0.16.0/nats/struct.Connection.html#method.request_multi)

**Comment by mmmries** [comment]
Thanks @autodidaddict for the suggestion. I'm not familiar with Rust, can I ask you a few questions about how that works and use that as a guide for how to add this to nats.ex?

* I didn't see how long the `request_multi` method was going to wait for answers? Does it just remain open until some other piece of code kills the subscription?
* Should we implement this as a synchronous function in Elixir? ie `Gnat.request_multi(conn, topic, request_message) # => [reply1, reply2, reply3]`?
  * we could also make it asynchronous and let the user just receive messages with the replies on their own?

If there are any other examples of how this kind of scatter-gather pattern is implemented I'd love to see those to help inform how it could be built here.

**Comment by autodidaddict** [comment]
the Rust version is nearly inscrutable because it returns a "future iterator" , which then expects the client to do something like time out. 
It would be ideal if this was a synchronous function where I could specify a maximum time to wait to gather the results and then just get `{:ok, [...]}` or `{:error, ...}` 

The implementation could work much the same way as it does in Rust (though blissfully without futures / async await syntax):
* provision a new mailbox and subscribe to it
* publish the consumer's request
* gather results until the timeout lapses
* cancel the subscription
* return the results as a list of messages

Ironically, we (the library consumers) have almost all the tools we need to do this on our own. the only thing we can't do is make a new mailbox because that functionality is (justifiably) private.

**Comment by mmmries** [comment]
Do you think we should also support a max number of returned messages? I'm thinking of something like:

```elixir
{:ok, gnat} = Gnat.start_link(connection_settings)
# gnat could also be a named connection that you started elsewhere in your supervision tree
Gnat.request_multi(gnat, "request_topic", request_message, max_milliseconds: 5_000, max_messages: 10)
# => {:ok, messages} or {:error, reason}
```

We could re-use the request inbox that is randomly generated when the connection is started, we subscribe to it as a wildcard so we can keep it open between requests, but add a unique suffix for each request. That helps out on performance if we don't have setup and cleanup another subscription with the server.

**Comment by mmmries** [comment]
The idea of the `max_messages` is to return early if we have already received all the messages we need to wait for. Maybe in some cases you know that in advance, but if you don't specify it, we would just wait 5 seconds without a limit

**Comment by autodidaddict** [comment]
Yeah I was thinking about that but thought it might be too much to ask for :) This is an ideal second use case for this function

**Comment by mmmries** [comment]
Got this added in and released as version `1.4.0` on hex 🎉
